File: app/src/main/java/soup/movie/ui/main/now/NowPresenter.java
package soup.movie.ui.main.now;

import com.jakewharton.rxrelay2.BehaviorRelay;

import javax.inject.Inject;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import soup.movie.data.MoobRepository;
import soup.movie.data.request.NowMovieRequest;
import soup.movie.data.response.NowMovieResponse;
import soup.movie.di.scope.FragmentScoped;
import soup.movie.ui.BasePresenter;

@FragmentScoped
public class NowPresenter extends BasePresenter<NowContract.View>
        implements NowContract.Presenter {

    private final MoobRepository moobRepository;

    private BehaviorRelay<Boolean> refreshRelay = BehaviorRelay.createDefault(true);

    @Inject
    NowPresenter(MoobRepository moobRepository) {
        this.moobRepository = moobRepository;
    }

    @Override
    protected void initObservable(CompositeDisposable subscriptions) {
        subscriptions.add(getViewStateObservable()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(view::render));
    }

    private Flowable<NowViewState> getViewStateObservable() {
        return refreshRelay.toFlowable(BackpressureStrategy.LATEST)
                .flatMap(ignore -> moobRepository.getNowList(NowMovieRequest.INSTANCE)
                        .map(NowMovieResponse::getList)
                        .map(NowViewState.DoneState::new)
                        .toFlowable());
    }

    @Override
    public void refresh() {
        refreshRelay.accept(true);
    }
}

File: app/src/main/java/soup/movie/ui/main/now/NowPresenter.kt
package soup.movie.ui.main.now

import com.jakewharton.rxrelay2.BehaviorRelay
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.internal.disposables.DisposableContainer
import soup.movie.data.MoobRepository
import soup.movie.data.request.NowMovieRequest
import soup.movie.ui.BasePresenter
import soup.movie.ui.main.now.NowContract.Presenter
import soup.movie.ui.main.now.NowContract.View
import soup.movie.ui.main.now.NowViewState.DoneState

class NowPresenter(private val moobRepository: MoobRepository) :
        BasePresenter<View>(), Presenter {

    private val refreshRelay = BehaviorRelay.createDefault(Unit)

    override fun initObservable(disposable: DisposableContainer) {
        super.initObservable(disposable)
        disposable.add(refreshRelay
                .switchMap { moobRepository.getNowList(NowMovieRequest) }
                .map { DoneState(it.list) }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { view?.render(it) })
    }

    override fun refresh() {
        refreshRelay.accept(Unit)
    }
}
